actors/builtin/storage_miner/storage_miner_actor.go

```diff
@@ -275,27 +271,35 @@ func (a *StorageMinerActor) PreCommitSector(rt Runtime, params *PreCommitSectorP
 	var st StorageMinerActorState
 	rt.State().Readonly(&st)
 	rt.ValidateImmediateCallerIs(st.Info.Worker)
-	if _, found := st.Sectors[params.info.SectorNumber]; found {
-		rt.Abort(exitcode.ErrIllegalArgument, "Sector number already exists in table")
+
+	store := adt.AsStore(rt)
+	if found, err := st.hasSectorNo(store, params.info.SectorNumber); err != nil {
+		rt.Abort(exitcode.ErrIllegalState, "failed to check sector %v: %v", params.info.SectorNumber, err)
+	} else if found {
+		rt.Abort(exitcode.ErrIllegalArgument, "sector %v already committed", params.info.SectorNumber)
 	}
 
+
 	cidx := rt.CurrIndices()
-	depositReq := cidx.StorageMining_PreCommitDeposit(st.Info.SectorSize, params.info.Expiration)
-	builtin.RT_ConfirmFundsReceiptOrAbort_RefundRemainder(rt, depositReq)
+	depositReq := cidx.StorageMining_PreCommitDeposit(st.getSectorSize(), params.info.Expiration)
+	confirmFundsReceiptOrAbort_RefundRemainder(rt, depositReq)
 
 	// TODO HS Check on valid SealEpoch
 
 	rt.State().Transaction(&st, func() interface{} {
-		st.PreCommittedSectors[params.info.SectorNumber] = SectorPreCommitOnChainInfo{
+		err := st.putPrecommittedSector(params.info.SectorNumber, &SectorPreCommitOnChainInfo{
 			Info:             params.info,
 			PreCommitDeposit: depositReq,
 			PreCommitEpoch:   rt.CurrEpoch(),
+		})
+		if err != nil {
+			rt.Abort(exitcode.ErrIllegalState, "failed to write pre-committed sector %v: %v", params.info.SectorNumber, err)
 		}
 		return nil
 	})
 
 	if params.info.Expiration <= rt.CurrEpoch() {
-		rt.Abort(exitcode.ErrIllegalArgument, "PreCommit sector must expire (%v) after now (%v)", params.info.Expiration, rt.CurrEpoch())
+		rt.Abort(exitcode.ErrIllegalArgument, "sector must expire (%v) after now (%v)", params.info.Expiration, rt.CurrEpoch())
```

Comment:
Is `params.info.Expiration` the epoch when sector expires or something different? 

This error message currently reads like: `sector must expire (35) after now (10)`, which makes me thing that the sector expires at epoch `45` - is this the intent or does it expire at epoch `35`?